src/haywire/http_server.c

```diff
@@ -378,6 +385,132 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
     }
 }
 
+void http_stream_on_read_pico(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf)
+{
+    http_connection* connection = (http_connection*)tcp->data;
+
+    if (nread > 0)
+    {
+        // TODO: Should this be at a higher level? Both parsers call this the same way.
+        http_request_buffer_consume(connection->buffer, nread);
+
+        printf("START\tNREAD: %d BUFUSED: %d PREVBUFLEN: %d\n",
+               nread,
+               http_request_buffer_get_used(connection->buffer),
+               connection->prevbuflen);
+        //fwrite(http_request_buffer_get_buffer(connection->buffer),
+        //       http_request_buffer_get_used(connection->buffer), 1, stdout);
+
+        while (connection->prevbuflen < http_request_buffer_get_used(connection->buffer))
+        {
+            printf("LOOP\tNREAD: %d BUFUSED: %d PREVBUFLEN: %d\n",
+                   nread,
+                   http_request_buffer_get_used(connection->buffer),
+                   connection->prevbuflen);
+
+            int parsed;
+            char buf[4096], *method, *path;
+            int minor_version;
+            struct phr_header headers[100];
+            size_t method_len;
+            size_t path_len;
+            size_t num_headers;
+            num_headers = sizeof(headers) / sizeof(headers[0]);
+
+            parsed = phr_parse_request(&http_request_buffer_get_buffer(connection->buffer)[connection->prevbuflen],
+                                       http_request_buffer_get_used(connection->buffer),
```

Comment:
Another thing worth mentioning is that whenever the URL, headers or body are discovered by Pico, you'll have to pin the underlying memory region like in https://github.com/haywire/haywire/blob/master/src/haywire/http_request.c#L173, so that if the request buffer is reallocated to a different memory region you can still recover those bits based on their relative position of the buffer, in the new buffer.